### PR TITLE
chore!: remove unsafe `std::env::set_var`

### PR DESCRIPTION
**Pull request by dj8yfo**
api example for build scripts - [build.rs](https://github.com/dj8yfo/factory-rust/blob/c5a970b4fb9436d2d9009035edbd1017e7f31a87/factory/build.rs#L9-L42) 

this has been extensively tested on following cases: 

## sample-crate

1.  https://github.com/dj8yfo/sample_no_workspace/tree/test_remove_unsafe_env_set_var-no-docker  => sample-crate-104.testnet 

2.  https://github.com/dj8yfo/sample_no_workspace/tree/32b06ab1be4bfd49cf54e6cda4e9d8481215195b  => sample-crate-105.testnet  

## factory-rust

1. ok (same bs / image) https://github.com/dj8yfo/factory-rust/tree/c5a970b4fb9436d2d9009035edbd1017e7f31a87  => repro-fct-76.testnet , ...

2. fail (new bs / old image) https://github.com/dj8yfo/factory-rust/tree/f3b9c74d4d759c32274554d598c5c91310e0bf47 =>  repro-fct-77.testnet , ...
fails reproducibility verification, across breaking change boundary (builder version `cargo-near-build 0.2.0` -> builder environment version `0.6.4`)

3. fail (old bs / new image) https://github.com/dj8yfo/factory-rust/tree/386f8ce9e0d7c8afcc931afb6dc64d564870adaf  =>  repro-fct-78.testnet , ...
fails reproducibility verification, across breaking change boundary (builder version `cargo-near-build 0.1.0` -> builder environment version `cargo-near-build 0.2.0`)

## neardevhub

1. https://github.com/dj8yfo/neardevhub-contract/tree/e1882a5149168c38841205ecc8370529bf2d3ecb => community.devhub-unse.testnet

2. https://github.com/dj8yfo/neardevhub-contract/tree/e1882a5149168c38841205ecc8370529bf2d3ecb => community.devhub-unsf.testnet

### CODE REVIEW

**Comment by dj8yfo** [comment]
@race-of-sloths ping

**Comment by race-of-sloths** [comment]
@dj8yfo Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!


[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=near/cargo-near/228&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/dj8yfo?type=bot&pr=near/cargo-near/228&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/dj8yfo)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/dj8yfo)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

File: cargo-near-build/src/types/near/build/input/mod.rs [hunk 1]
```diff
@@ -47,6 +47,13 @@ pub struct Opts {
     /// additional environment key-value pairs, that should be passed to underlying
     /// build commands
     pub env: Vec<(String, String)>,
+    /// additional environment key-value pairs, that should be passed to underlying
+    /// build commands
+    ///
+    /// unlike [Self::env] this env is *mute*: it's not reflected in `--env` flag-components of environment
+    /// [crate::env_keys::nep330::BUILD_COMMAND] variable,
+    /// set during build
+    pub mute_env: Vec<(String, String)>,
```
Comment (frol) [inline cargo-near-build/src/types/near/build/input/mod.rs:56]: I find this naming unintuitive. My suggestions:

* nep330_env
* near_env ("near" is a regular word and makes it weird as well)
* implicit_env
* cargo_near_env

Also, would it be harmful if we pass these env variables via `env` as well instead of creating a separate flow for them?

File: cargo-near-build/src/types/cargo/metadata.rs [hunk 2]
```diff
@@ -70,12 +77,17 @@ impl CrateMetadata {
...
+    cargo_target_dir: Option<&buildtime_env::CargoTargetDir>,
 ) -> eyre::Result<(cargo_metadata::Metadata, Package)> {
     tracing::info!("Fetching cargo metadata for {}", manifest_path.path);
     let mut cmd = MetadataCommand::new();
     if !no_locked {
         cmd.other_options(["--locked".to_string()]);
     }
+    if let Some(target_dir) = cargo_target_dir {
+        let (key, value) = target_dir.entry();
+        cmd.env(key, value);
```
Comment (dj8yfo) [inline cargo-near-build/src/types/cargo/metadata.rs:89]: this helps resolve output destination to `env!(CARGO_TARGET_DIR)/near/<package>/<package>.wasm`
similarly to what would've been done in case of `std::env::set_var` before building

File: cargo-near-build/src/near/build/mod.rs [hunk 3]
```diff
@@ -15,18 +16,26 @@ use crate::{
...
 
-pub mod export;
-
 /// builds a contract whose crate root is current workdir, or identified by [`Cargo.toml`/BuildOpts::manifest_path](crate::BuildOpts::manifest_path) location
-pub fn run(args: Opts) -> eyre::Result<CompilationArtifact> {
-    VersionMismatch::export_builder_and_near_abi_versions();
-    export::nep_330_build_command(&args)?;
+pub fn run(
+    args: Opts,
+    implicit_env_opts: Option<input::implicit_env::Opts>,
```
Comment (frol) [inline cargo-near-build/src/near/build/mod.rs:31]: Please, look at the public API from the usage point of view.

1. This (Opts, Option<Opts>) is really hard to follow.
2. `nep330_contract_path` doesn't make sense to me as a user of the package - I can understand `contract_path`, but I would never thought about it in terms of "Oh, I need to specify NEP330 parameters, here we go nep330_contract_path".

File: cargo-near-build/src/near/build_extended/mod.rs [hunk 4]
```diff
@@ -1,54 +1,52 @@
...
 use rustc_version::Version;
 
+use crate::extended::BuildScriptOpts;
+
+/// only single-threaded build-scripts are supported
+///
+/// this function cannot be run concurrently with itself, as it changes working dir
+/// of process, executing it.
+/// changing current working dir concurrently from different threads
+/// may entail incorrect results
```
Comment (frol) [inline cargo-near-build/src/near/build_extended/mod.rs:15]: Can we resolve this limitation? E.g. specify it as an explicit command line argument to `cargo build`, or [set it as part of `exec`](https://doc.rust-lang.org/stable/std/process/struct.Command.html#method.current_dir). I believe that the current implementation may cause problems with tests and other concurrent runs that will be hard to debug.

**Review by frol — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -39,6 +39,7 @@ jobs:
         run: | 
           cargo check -p cargo-near
           cargo check -p cargo-near-build --all-features
+          cargo test -p cargo-near-build --all-features --doc
```
Comment (dj8yfo) [inline .github/workflows/test.yml:42]: https://github.com/near/cargo-near/actions/runs/11331185786/job/31510648632

File: Cargo.lock [hunk 6]
```diff
@@ -1450,21 +1452,21 @@ dependencies = [
...
 name = "futures-task"
-version = "0.3.30"
+version = "0.3.31"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "38d84fa142264698cdce1a9f9172cf383a0c82de1bddcf3092901442c4097004"
+checksum = "f90f7dce0722e95104fcb095585910c0977252f286e354b5e3bd38902cd99988"
 
 [[package]]
 name = "futures-util"
-version = "0.3.30"
```
Comment (dj8yfo) [inline Cargo.lock:1465]: 0.3.30 has been yanked, `cargo update -p futures-util`

File: cargo-near-build/Cargo.toml [hunk 7]
```diff
@@ -45,10 +46,15 @@ dist = false
 [features]
 default = []
 build_script = []
-abi_build = [] 
+abi_build = []
 docker = [
-  "dep:url", "dep:serde", "dep:git2", 
-  "dep:home", "dep:pathdiff", "dep:unix_path", 
-  "dep:tempfile", "dep:nix", "dep:shell-words"
+  "dep:url",
```
Comment (dj8yfo) [inline cargo-near-build/Cargo.toml:51]: fmt applied with https://taplo.tamasfe.dev/

**Review by frol — APPROVED** [review]
@race-of-sloths score 13
